kernel/src/process/signal/mod.rs

```diff
@@ -197,23 +200,49 @@ pub fn handle_user_signal(
     let siginfo_addr = stack_pointer;
 
     // 2. Write ucontext_t.
-    stack_pointer = alloc_aligned_in_user_stack(stack_pointer, mem::size_of::<ucontext_t>(), 16)?;
+
+    // Store FPU context on stack
+    let fpu_context = ctx.thread_local.fpu().clone_context();
+    let fpu_context_addr = alloc_aligned_in_user_stack(
+        stack_pointer,
+        fpu_context.as_bytes().len(),
+        align_of::<FpuContext>(),
+    )?;
+    let mut fpu_context_reader = VmReader::from(fpu_context.as_bytes());
+    user_space.write_bytes(fpu_context_addr as _, &mut fpu_context_reader)?;
+    // Reset FPU context
+    ctx.thread_local.fpu().set_context(FpuContext::new());
+
+    stack_pointer =
+        alloc_aligned_in_user_stack(fpu_context_addr, mem::size_of::<ucontext_t>(), 16)?;
     let mut ucontext = ucontext_t {
         uc_sigmask: mask.into(),
         ..Default::default()
     };
-    ucontext
-        .uc_mcontext
-        .inner
-        .gp_regs
-        .copy_from_raw(user_ctx.general_regs());
+    ucontext.uc_mcontext.copy_user_regs_from(user_ctx);
     let sig_context = ctx.thread_local.sig_context().get();
     if let Some(sig_context_addr) = sig_context {
         ucontext.uc_link = sig_context_addr;
     } else {
         ucontext.uc_link = 0;
     }
-    // TODO: store fp regs in ucontext
+    cfg_if::cfg_if! {
+        if #[cfg(target_arch = "x86_64")] {
+            ucontext
+                .uc_mcontext
+                .set_fpu_context_addr(fpu_context_addr as _);
+
+            const UC_FP_XSTATE: u64 = 1 << 0;
+            ucontext.uc_flags = UC_FP_XSTATE;
+        } else if #[cfg(target_arch = "riscv64")] {
+            // No FPU support in RISC-V, so we do nothing here. However, an
+            // assertion is added to ensure that we'll be notified when we add
+            // the FPU support someday.
+            assert_eq!(fpu_context.as_bytes().len(), 0);
+        } else {
+             compile_error!("unsupported target");
```

Comment:
Rust format does not seem to work here?
```suggestion
            compile_error!("unsupported target");
```